students/johnn/session09/mailroom.py

```diff
@@ -0,0 +1,688 @@
+            self.middle_name = middle_name.title()
+            self.suffix = suffix.upper()
+
+        if (donations != None) and (donation != None):
+            raise ValueError("You cannot set 'donations' and 'donation' at the same time.")
+
+        if donation != None:
+            # if a donation is passed in, initialize the donations, then add the donation
+            self._donations = list()
+            self.add_donation(donation)
+        else:
+            if donations != None:
+                # TODO donations need more validation
+                self._donations = donations
+            else:
+                self._donations = list()
+            
+    @property
+    def id(self):
+        """ return the donor id """
+        return self._id
+
+    @id.setter
+    def id(self, value):
+        """ donor id cannot be changed once established """
+        raise AttributeError("You cannot change the 'id' after the record has been created.")
+
+    @property
+    def donations(self):
+        """ print all donations """
+        return self._donations
+
+    @donations.setter
+    def donations(self, value):
+        """ allow bulk setting of donation entries """
+        self._donations = value
+
+    def add_donation(self, value):
+        """ add a single donation """
+        today = datetime.datetime.utcnow().date().isoformat() + "Z"
+        try:
+            value = float(value)
+        except ValueError:
+            raise ValueError("Donations must be values.")
+        self._donations.append( { "amount": value, "date": today } )
+
+    @property
+    def number_donations(self):
+        """ return the number of donations """
+        return len(self._donations)
+
+    @property
+    def total_donations(self):
+        """ return a sum of the donations """
+        total_donations = 0
+        for donation in self.donations:
+            total_donations += donation["amount"]
+        return round(total_donations, 2)
+
+    @property
+    def average_donations(self):
+        """ return the average donation amount """
+        try:
+            return round(self.total_donations / self.number_donations, 2)
+        except ZeroDivisionError:
+            return 0
+
+    @property
+    def first_name(self):
+        """ return the first name """
+        return self._first_name
+
+    @first_name.setter
+    def first_name(self, value):
+        """ set the first name """
+        self._first_name = value.title()
+
+    @property
+    def middle_name(self):
+        """ return the middle name """
+        return self._middle_name
+
+    @middle_name.setter
+    def middle_name(self, value):
+        """ set the middle name """
+        self._middle_name = value.title()
+
+    @property
+    def last_name(self):
+        """ return the last name """
+        return self._last_name
+
+    @last_name.setter
+    def last_name(self, value):
+        """ set the last name """
+        self._last_name = value.title()
```

Comment:
I'm not sure I'd bother with all this -- but it is a pretty nifty way to normalize the names.

I can't help thinking there would be a way to avoid the duplication of code here -- try to remember to bring this up when we talk about metaprogramming next quarter.